File: lecture_010(RecursionwithArraylists)/l_003_getstairspath.cpp
#include<iostream>
#include<vector>
#include<string>
using namespace std;

vector<string> getstairspath(int n) {
    if(n==0) {
        vector<string> bres;
        bres.push_back("");
        return bres;
    }
    else if(n<0) {
        vector<string> bres;
        return bres;
    }
    vector<string> paths1 = getstairspath(n-1);
    vector<string> paths2 = getstairspath(n-2);
    vector<string> paths3 = getstairspath(n-3);

    vector<string> paths;
    for(string path: paths1) {
        paths.push_back("1" + path);
    }
    for(string path: paths2) {
        paths.push_back("2" + path);
    }
    for(string path: paths3) {
        paths.push_back("3" + path);
    }
    return paths;
}

int main() {
    int n;
    cin>>n;
    vector<string> result = getstairspath(n);
    cout<<"[";
    for(int i=0; i<result.size()-1; i++) {
        cout<<result[i]<<", ";
    }
    cout<<result[result.size()-1]<<"]";

    return 0;
}

File: lecture_010(RecursionwithArraylists)/l_003_getstairspath.java
import java.util.Scanner;
import java.util.ArrayList;
public class l_003_getstairspath {
    public static Scanner scn = new Scanner(System.in);
    public static void main(String[] args) {
        int n = scn.nextInt();
        ArrayList<String> result = getStairsPath(n);
        System.out.println(result); 
    }
    public static ArrayList<String> getStairsPath(int n) {
        if(n==0) {
            ArrayList<String> bres = new ArrayList<>();
            bres.add("");
            return bres;
        }
        else if(n<0) {
            ArrayList<String> bres = new ArrayList<>();
            return bres;
        }
        ArrayList<String> paths1 = getStairsPath(n-1);
        ArrayList<String> paths2 = getStairsPath(n-2);
        ArrayList<String> paths3 = getStairsPath(n-3);

        ArrayList<String> paths = new ArrayList<>();
        for(String path: paths1) {
            paths.add(1 + path);
        }
        for(String path: paths2) {
            paths.add(2 + path);
        }
        for(String path: paths3) {
            paths.add(3 + path);
        }
        return paths;
    }
}
